<div style="background-color: #323232; color: white; padding: 10px; text-align: center;">
    <h1> <strong>LAB-03: ỨNG DỤNG VÀ CÔNG CỤ TRỰC QUAN HÓA DỮ LIỆU</strong> </h1>
</div>

<h1 style="color: black;"><strong>Thông tin nhóm</strong></h1>

<table style="float: left;">
    <tr>
        <td style="text-align: center;"><strong>MSSV</strong></td>
        <td style="text-align: center;"><strong>Họ và Tên</strong></td>
        <td style="text-align: center;"><strong>Email</strong></td>
    </tr>
    <tr>
        <td style="text-align: left;">20120270</td>
        <td style="text-align: left;">Cao Tấn Đức</td>
        <td style="text-align: left;">20120270@student.hcmus.edu.vn</td>
    </tr>
    <tr>
        <td style="text-align: left;">20120284</td>
        <td style="text-align: left;">Lê Đức Hậu</td>
        <td style="text-align: left;">20120284@student.hcmus.edu.vn</td>
    </tr>
    <tr>
        <td style="text-align: left;">20120288</td>
        <td style="text-align: left;">Nguyễn Trung Hiếu</td>
        <td style="text-align: left;">20120288@student.hcmus.edu.vn</td>
    </tr>
    <tr>
        <td style="text-align: left;">20120305</td>
        <td style="text-align: left;">Võ Thị Kiều Khanh</td>
        <td style="text-align: left;">20120305@student.hcmus.edu.vn</td>
    </tr>
</table>


<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
           letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b> | Tổng quan về bộ dữ liệu </b></p>
</div>

Dữ liệu được nhóm chúng em chọn liên quan đến Cryptocurrency – Tiền điện tử, cụ thể là đồng Bitcoin (BTC). 

Dữ liệu được lấy từ [kaggle.com](https://kaggle.com/) có tên [Bitcoin Historical Data](https://www.kaggle.com/datasets/mczielinski/bitcoin-historical-data), chứa dữ liệu về giá của Bitcoin trong khoảng thời gian 1 phút từ các sàn giao dịch được chọn lọc, bắt đầu từ tháng 1 năm 2012 đến hết tháng 3 năm 2021. 

Dataset này chứa <strong style="color: green;">4857377</strong> dòng dữ liệu giao dịch của Bitcoin trên sàn **Binance** bao gồm <strong style="color: green;">8</strong> trường dữ liệu


<table style="float: left;">
    <tr>
        <td style="text-align: left;"><strong>Tên cột</strong></td>
        <td style="text-align: left;"><strong>Ý nghĩa</strong></td>
    </tr>
    <tr>
        <td style="text-align: left;">Timestamp</td>
        <td style="text-align: left;">Dấu thời gian Unix (số giây kể từ ngày 1 tháng 1 năm 1970) của thời điểm bắt đầu khoảng thời gian 1 phút</td>
    </tr>
    <tr>
        <td style="text-align: left;">Open</td>
        <td style="text-align: left;">Giá của Bitcoin khi bắt đầu khoảng thời gian 1 phút</td>
    </tr>
    <tr>
        <td style="text-align: left;">High</td>
        <td style="text-align: left;">Giá Bitcoin cao nhất trong khoảng thời gian một phút</td>
    </tr>
    <tr>
        <td style="text-align: left;">Low</td>
        <td style="text-align: left;">Giá thấp nhất của Bitcoin trong khoảng thời gian 1 phút</td>
    </tr>
    <tr>
        <td style="text-align: left;">Close</td>
        <td style="text-align: left;">Giá của Bitcoin khi kết thúc khoảng thời gian 1 phút</td>
    </tr>
    <tr>
        <td style="text-align: left;">Volume_(BTC)</td>
        <td style="text-align: left;">Tổng khối lượng giao dịch của Bitcoin trong khoảng thời gian một phút, tính bằng BTC</td>
    </tr>
    <tr>
        <td style="text-align: left;">Volume_(Currency)</td>
        <td style="text-align: left;">Tổng khối lượng giao dịch của Bitcoin trong khoảng thời gian 1 phút, tính bằng USD</td>
    </tr>
    <tr>
        <td style="text-align: left;">Wighted_Price</td>
        <td style="text-align: left;">Volume Weight Average Price - Giá trung bình theo khối lượng của Bitcoin trong khoảng thời gian 1 phút, được tính bằng công thức Weighted_Price = Volume_(Currency) / Volume_(BTC)</td>
    </tr>
</table>

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
           letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b> | Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp hay không? </b></p>
</div>

Cột có kiểu dữ liệu là int64: `Timestamp` 

Các cột có kiểu dữ liệu là float64: `Open`, `High`, `Low`, `Close`, `Volume_(BTC)`, `Volume_(Currency)`, `Weighted_Price` 

Cột `Timestamp` có kiểu dữ liệu chưa phù hợp: 

- Việc sử dụng Unix-time có thể dẫn đến khó hoặc không thể hiểu rõ ràng đối với những người sử dụng dữ liệu với mục đích tham khảo nhanh 

- Đối với dữ liệu time-series để có thể dễ dàng rút trích được ý nghĩa cũng như là để nhận xét được từ những  biểu đồ tạo ra bằng các phương pháp trực quan, quá trình trực quan cần dữ liệu thể hiện ngày tháng rõ ràng. 

$\Rightarrow$ Đề xuất xử lý và chuyển đổi thời gian từ Unix-time sang dạng datetime64 thông thường. 

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b> | Tiền xử lý  </b></p>
</div>

1. `Timestamp`: chuyển từ unix-time sang dateime 

2. Những dòng bị thiếu dữ liệu có các  cách xử lý: 

	+ Xóa những dòng này nhưng không đảm bảo trình tự logic của thời gian. 

	+ Điền vào những vị trí bị thiếu: 

	 	* Forward fill 

		* Backward fill 

$\Rightarrow$ Hướng giải quyết của nhóm: xóa những dòng bị thiếu

3. Sau khi xoá những dòng dữ liệu bị thiếu, thông tin về giá theo từng phút sẽ không đảm bảo tính liên tục

$\Rightarrow$ Gom nhóm theo ngày để đảm bảo tính liên tục của bộ dữ liệu


In [1]:
# Import library
import pandas as pd
import datetime

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Load dataset
df = pd.read_csv('./bitstampUSD_1-min_data_2012-01-01_to_2021-03-31.csv')
df.head()

Timestamp  Open  High   Low  Close  Volume_(BTC)  Volume_(Currency)  \
0  1325317920  4.39  4.39  4.39   4.39      0.455581                2.0   
1  1325317980   NaN   NaN   NaN    NaN           NaN                NaN   
2  1325318040   NaN   NaN   NaN    NaN           NaN                NaN   
3  1325318100   NaN   NaN   NaN    NaN           NaN                NaN   
4  1325318160   NaN   NaN   NaN    NaN           NaN                NaN   

   Weighted_Price  
0            4.39  
1             NaN  
2             NaN  
3             NaN  
4             NaN

Cột `Timestamp` có kiểu dữ liệu lưu trữ là `int64`, chứa dữ liệu về thời gian kiểu Unix-time $\rightarrow$ Để giúp việc đọc và trực quan hoá dễ dàng hơn, cần chuyển cột này sang dạng `datetime`

In [3]:
# Unix-time to datetime
df.Timestamp = pd.to_datetime(df.Timestamp, unit='s')

### Kiểm tra missing value

In [4]:
# Missing value in dataset
df.isnull().sum()

Timestamp                  0
Open                 1243608
High                 1243608
Low                  1243608
Close                1243608
Volume_(BTC)         1243608
Volume_(Currency)    1243608
Weighted_Price       1243608
dtype: int64

In [5]:
# Show all row with at least 1 missing value
nan_rows = df[df.isnull().any(axis=1)]
nan_rows

Timestamp  Open  High  Low  Close  Volume_(BTC)  \
1       2011-12-31 07:53:00   NaN   NaN  NaN    NaN           NaN   
2       2011-12-31 07:54:00   NaN   NaN  NaN    NaN           NaN   
3       2011-12-31 07:55:00   NaN   NaN  NaN    NaN           NaN   
4       2011-12-31 07:56:00   NaN   NaN  NaN    NaN           NaN   
5       2011-12-31 07:57:00   NaN   NaN  NaN    NaN           NaN   
...                     ...   ...   ...  ...    ...           ...   
4856245 2021-03-30 05:09:00   NaN   NaN  NaN    NaN           NaN   
4856308 2021-03-30 06:12:00   NaN   NaN  NaN    NaN           NaN   
4856376 2021-03-30 07:20:00   NaN   NaN  NaN    NaN           NaN   
4857340 2021-03-30 23:24:00   NaN   NaN  NaN    NaN           NaN   
4857368 2021-03-30 23:52:00   NaN   NaN  NaN    NaN           NaN   

         Volume_(Currency)  Weighted_Price  
1                      NaN             NaN  
2                      NaN             NaN  
3                      NaN             NaN  
4                      NaN             NaN  
5                      NaN             NaN  
...                    ...             ...  
4856245                NaN             NaN  
4856308                NaN             NaN  
4856376                NaN             NaN  
4857340                NaN             NaN  
4857368                NaN             NaN  

[1243608 rows x 8 columns]

### Nhận xét:
+ Có **1243608** dòng đều bị thiếu dữ liệu ở các cột `['Open', 'High', 'Low', 'Close', 'Volume_(BTC)', 'Volume_(Currency)', 'Weighted_Price']`
+ Các dòng thiếu dữ liệu này thường là một vài phút hay vài giờ trong 1 ngày $\rightarrow$ Có thể xoá các dòng bị thiếu dữ liệu, nếu như không có ngày nào thiếu dữ liệu toàn bộ 24h

In [6]:
nan_rows['day'] = nan_rows['Timestamp'].dt.date
# Count missing values by day
nan_rows_by_day = nan_rows.groupby('day')['Timestamp'].count().reset_index(name='count')

nan_rows_by_day.head()

day  count
0  2011-12-31    964
1  2012-01-01   1437
2  2012-01-02   1439
3  2012-01-03   1432
4  2012-01-04   1431

In [7]:
# Find day have 1440 missing values (1 day = 1440 min)
nan_rows_by_day[nan_rows_by_day['count'] == 1440].head()

Empty DataFrame
Columns: [day, count]
Index: []

### Nhận xét:
+ Không có bất kỳ ngày nào bị thiếu toàn bộ dữ liệu $\rightarrow$ Có thể xoá tất cả các dòng thiếu dữ liệu

In [8]:
# Drop NaN
df.dropna(axis=0, how='any', inplace=True)

### Tạo bảng gom nhóm dataset theo ngày

In [9]:
df_day = df.copy()
df_day['Timestamp'] = df_day['Timestamp'].dt.tz_localize(None)
df_day = df_day.groupby([pd.Grouper(key='Timestamp', freq='D')]).agg({
    'Open': 'first',
    'High': 'max',
    'Low': 'min',
    'Close': 'last',
    'Volume_(BTC)': 'sum',
    'Volume_(Currency)': 'sum',
    'Weighted_Price': 'mean'
}).reset_index()
df_day = df_day.set_index('Timestamp')
df_day.head()

Open  High   Low  Close  Volume_(BTC)  Volume_(Currency)  \
Timestamp                                                              
2011-12-31  4.39  4.58  4.39   4.58     95.317878         425.320338   
2012-01-01  4.58  5.00  4.58   5.00     21.602000         105.779160   
2012-01-02  5.00  5.00  5.00   5.00     19.048000          95.240000   
2012-01-03  5.32  5.32  5.14   5.29     88.037281         464.805210   
2012-01-04  4.93  5.57  4.93   5.57    107.233260         568.076197   

            Weighted_Price  
Timestamp                   
2011-12-31        4.471603  
2012-01-01        4.806667  
2012-01-02        5.000000  
2012-01-03        5.252500  
2012-01-04        5.208159

### Xuất dataset theo ngày ra file "BTC-Historical-Price.csv"

In [10]:
df_day.to_csv('BTC-Historical-Price.csv')